File: script/phys/ammo/RigidBody.ts
import { MotionState } from "./MotionState.js";
import { Shape } from "./Shape.js";
import { Transform } from "./Transform.js";
import { impl as ammo } from "./context.js";

const bt_vec3_a = new ammo.btVector3(0, 0, 0);
const bt_transform_a = new ammo.btTransform();

export class RigidBody {
    readonly impl: any;

    private _mass: number = 0;
    get mass(): number {
        return this._mass;
    }
    set mass(value: number) {
        const shape = ammo.castObject(this.impl.getCollisionShape(), ammo.btCompoundShape);
        bt_vec3_a.setValue(0, 0, 0);
        shape.calculateLocalInertia(value, bt_vec3_a);
        this.impl.setMassProps(value, bt_vec3_a);

        this._mass = value;
    }

    private _worldTransform = new Transform(null);
    get worldTransform(): Transform {
        this._worldTransform.impl = this.impl.getWorldTransform();
        return this._worldTransform;
    }

    set worldTransform(transform: Transform) {
        this.impl.setWorldTransform(transform.impl);
    }

    constructor(motionState: MotionState) {
        const info = new ammo.btRigidBodyConstructionInfo(1, motionState.impl, new ammo.btCompoundShape());
        this.impl = new ammo.btRigidBody(info);
        ammo.destroy(info);

        // bt_vec3_a.setValue(0, 0, 0);
        // this.impl.setMassProps(0, bt_vec3_a);
    }

    addShape(shape: Shape) {
        bt_transform_a.setIdentity();
        ammo.castObject(this.impl.getCollisionShape(), ammo.btCompoundShape).addChildShape(bt_transform_a, shape.impl);
    }

    updateShapeTransform(shape: Shape, transform: Transform) {
        let index = -1;
        const compound = ammo.castObject(this.impl.getCollisionShape(), ammo.btCompoundShape);
        for (let i = 0; i < compound.getNumChildShapes(); i++) {
            if (ammo.compare(compound.getChildShape(i), shape.impl)) {
                index = i;
            }
        }
        compound.updateChildTransform(index, transform.impl);
    }
}

File: puttyknife/phys/bullet-end/RigidBody.cpp
#include <phys/RigidBody.hpp>
#include <BulletDynamics/Dynamics/btRigidBody.h>
#include <BulletCollision/CollisionShapes/btCompoundShape.h>
#include <LinearMath/btTransform.h>

phys::RigidBody *physRigidBody_new()
{
    btCompoundShape *compound = new btCompoundShape();
    btRigidBody::btRigidBodyConstructionInfo info(1.0, nullptr, compound);

    return new btRigidBody(info);
}

phys::CollisionShape *physRigidBody_getCollisionShape(phys::RigidBody *body)
{
    return body->getCollisionShape();
}

void physRigidBody_setMassProps(phys::RigidBody *body, float mass, const phys::Vector3 *inertia)
{
    body->setMassProps(mass, *inertia);
}

const phys::Transform *physRigidBody_getWorldTransform(phys::RigidBody *body)
{
    return &body->getWorldTransform();
}

void physRigidBody_setWorldTransform(phys::RigidBody *body, const phys::Transform *t)
{
    body->setWorldTransform(*t);
}

void physRigidBody_addShape(phys::RigidBody *body, phys::CollisionShape *shape)
{
    btTransform t;
    t.setIdentity();
    reinterpret_cast<btCompoundShape *>(body->getCollisionShape())->addChildShape(t, shape);
}

void physRigidBody_updateShapeTransform(phys::RigidBody *body, phys::CollisionShape *shape, const phys::Transform *t)
{
    btCompoundShape *compound = reinterpret_cast<btCompoundShape *>(body->getCollisionShape());
    int num = compound->getNumChildShapes();
    int i;
    for (i = 0; i < num; i++)
    {
        if (shape == compound->getChildShape(i))
        {
            break;
        }
    }
    compound->updateChildTransform(i, *t);
}
